qiskit/circuit/quantumcircuit.py

```diff
@@ -1283,41 +1286,89 @@ def add_register(self, *regs: Union[Register, int, Sequence[Bit]]) -> None:
 
             if isinstance(register, QuantumRegister):
                 self.qregs.append(register)
-                new_bits = [bit for bit in register if bit not in self._qubit_set]
-                self._qubits.extend(new_bits)
-                self._qubit_set.update(new_bits)
+
+                for idx, bit in enumerate(register):
+                    if bit in self._qubit_indices:
+                        self._qubit_indices[bit][1].append((register, idx))
+                    else:
+                        self._qubits.append(bit)
+                        self._qubit_indices[bit] = BitLocations(
+                            len(self._qubits) - 1, [(register, idx)]
+                        )
+
             elif isinstance(register, ClassicalRegister):
                 self.cregs.append(register)
-                new_bits = [bit for bit in register if bit not in self._clbit_set]
-                self._clbits.extend(new_bits)
-                self._clbit_set.update(new_bits)
+
+                for idx, bit in enumerate(register):
+                    if bit in self._clbit_indices:
+                        self._clbit_indices[bit][1].append((register, idx))
+                    else:
+                        self._clbits.append(bit)
+                        self._clbit_indices[bit] = BitLocations(
+                            len(self._clbits) - 1, [(register, idx)]
+                        )
+
             elif isinstance(register, list):
                 self.add_bits(register)
             else:
                 raise CircuitError("expected a register")
 
     def add_bits(self, bits: Sequence[Bit]) -> None:
         """Add Bits to the circuit."""
-        duplicate_bits = set(self.qubits + self.clbits).intersection(bits)
+        duplicate_bits = set(self._qubit_indices).union(self._clbit_indices).intersection(bits)
         if duplicate_bits:
             raise CircuitError(f"Attempted to add bits found already in circuit: {duplicate_bits}")
 
         for bit in bits:
             if isinstance(bit, AncillaQubit):
                 self._ancillas.append(bit)
-
+                self._qubit_indices[bit] = BitLocations(len(self._ancillas) - 1, [])
```

Comment:
Is adding this line correct, here?  Previously, the ancillas were never added to `self._qubit_set`, but now we're adding them to its replacement `_qubit_indices`.  It does _seem_ more useful to me this way, but I'm just noting that it's a behavioural change from before.  Also, if we _don't_ have it, then the `isinstance(bit, Qubit)` protection on accessing `self._qubit_indices` would be incorrect, since `AncillaQubit` instances would also pass that test.

Reply:
Good catch, `AncillaQubits` should be stored in `_qubit_indices` according to their index in `self._qubits`, not `self._ancillas` so the pass through to the `isinstance(bit, Qubit)` block is sufficient.

Reply:
Looking again, I realise my original comment was slightly wrong. They _were_ added by the next part of the branch, but this way is now better.